### PR TITLE
Add < and ==, making polynomial an ordered ring.

### PR DESCRIPTION
**Pull request by jeremy-murphy**
Hey, just wondering if there is interest in improving the `polynomial` class?

I have a few things in mind but ordered ring operators were just an easy thing to start off with.

Cheers.

### CODE REVIEW

File: include/boost/math/tools/polynomial.hpp [hunk 1]
```diff
@@ -300,6 +307,18 @@ inline polynomial<T> operator * (const U& a, const polynomial<T>& b)
...
+bool operator == (polynomial<T> const &a, polynomial<T> const &b)
+{
+    return a.data() == b.data();
+}
+
+template <typename T>
+bool operator < (polynomial<T> const &a, polynomial<T> const &b)
+{
+    return a.data() < b.data();
+}
```
Comment (jeremy-murphy) [inline include/boost/math/tools/polynomial.hpp]: Hmmm, hold on, I just realized that maybe polynomial order is not as straightforward as lexicographical order. I will sleep on it for now.

**Comment by jzmaddock** [comment]
I don't think polynomials are really ordered?

**Comment by jeremy-murphy** [comment]
You're right, they don't have a natural total ordering like integers do, it's just a default total ordering for practical convenience. Originally I was just going to add equality, which is necessary for GCD, but then I gave in to the temptation of adding more since it seemed harmless.

**Comment by jeremy-murphy** [comment]
An argument in favour of granting a default total ordering is that it is required, strictly speaking, for a type to be *regular* by the definition in Elements of Programming, and thus interoperable with all the facilities of the STL. (E.g. it permits sort and binary search.)

**Comment by jeremy-murphy** [comment]
However, total ordering is not necessary for GCD, so if it is too contentious I am happy to take it out. Equality is the essential thing.

**Comment by jzmaddock** [comment]
I think making this play nice with std lib containers is a good argument in favour of these.

However,

* If this class is to get a major workover, then I'd rather make all the changes (or at least most of them) at once, and:
* I suspect these operators sort in the "wrong" order (so far as there is one) in that they sort on the x^0 coefficient first and the x^n coefficient last.

**Comment by jeremy-murphy** [comment]
OK, sure, that makes sense. I was just going on the usual assumption that smaller pull requests are easier to review but in this case it's preferable to see the whole impact in one go. 

Ah, you're right, of course. I agree there is no right or wrong order, only a convention. So the correct definition of `operator<` would be to swap the arguments, `return b.data() < a.data()`, right? Anyway, I'm going to leave ordering out for now because it is distracting from the real purpose of GCD.

I'll close this PR and come back soon with a complete one for Euclidean GCD, but I'm inclined to leave Stein GCD to a separate PR. Is that OK?

**Comment by jzmaddock** [comment]
On 31/10/2015 02:58, Jeremy W. Murphy wrote:
>
> OK, sure, that makes sense. I was just going on the usual assumption 
> that smaller pull requests are easier to review but in this case it's 
> preferable to see the whole impact in one go.
>
> Ah, you're right, of course. I agree there is no right or wrong order, 
> only a convention. So the correct definition of |operator<| would be 
> to swap the arguments, |return b.data() < a.data()|, right? Anyway, 
> I'm going to leave ordering out for now because it is distracting from 
> the real purpose of GCD.
>
> I'll close this PR and come back soon with a complete one for 
> Euclidean GCD, but I'm inclined to leave Stein GCD to a separate PR. 
> Is that OK?
>
OK that sounds fine, Best John.

**Pull request closed without merging**